### PR TITLE
feat: rewrite e2e test(test-e2e-route-with-management-fields) using ginkgo

### PR DESCRIPTION
**Pull request by bisakhmondal**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**Related issues**

#1500 

**Checklist:**

- [ ] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+			Method: http.MethodPut,
+			Body: `{
+					"name": "route1",
+					"uri": "/hello",
+					"name": "jack",
+					"desc": "config route with name and desc",
+					"upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_management_fields_test.go:53]: nodes host please use the base.UpstreamIp
```
"nodes": {
			"` + base.UpstreamIp + `:1980": 1
		}
```

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+					"uri": "/hello",
+					"labels": {
+						"build":"16",
+						"env":"production",
+						"version":"v2"
+					},
+					"upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_management_fields_test.go:272]: ditto

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+					"labels": {
+						"build":"17",
+						"env":"dev",
+						"version":"v2",
+						"extra": "test"
+					},
+					"upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_management_fields_test.go:297]: ditto

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+			Object: base.ManagerApiExpect(),
+			Path:   "/apisix/admin/routes/r1",
+			Method: http.MethodPut,
+			Body: `{
+				"name": "route1",
+				"uri": "/hello",
+				"create_time": 1608792721,
+				"upstream": {
+					"nodes": {
+						"172.16.238.20:1980": 1
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_management_fields_test.go:404]: ditto

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+			Path:   "/apisix/admin/routes/r1",
+			Method: http.MethodPut,
+			Body: `{
+				"name": "route1",
+				"uri": "/hello",
+				"create_time": 1608792721,
+				"update_time": 1608792721,
+				"upstream": {
+					"nodes": {
+						"172.16.238.20:1980": 1
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_management_fields_test.go:441]: ditto

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+					"uri": "/hello",
+					"labels": {
+						"build":"16",
+						"env":"production",
+						"version":"v2"
+					},
+					"upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_management_fields_test.go:207]: ditto

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+			Method: http.MethodPut,
+			Body: `{
+					"name": "route1",
+					"uri": "/hello",
+					"name": "new jack",
+					"desc": "new desc",
+					"upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_management_fields_test.go:138]: ditto

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+
+	ginkgo.It("delete the route (r1)", func() {
+		base.RunTestCase(base.HttpTestCase{
+			Object:       base.ManagerApiExpect(),
+			Method:       http.MethodDelete,
+			Path:         "/apisix/admin/routes/r1",
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+		})
+	})
```
Comment (Jaycean) [inline api/test/e2enew/route/route_with_management_fields_test.go:184]: After the route is deleted, we need to check whether the route is deleted.

Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go:180]: Done :heavy_check_mark:

File: api/test/e2enew/route/route_with_management_fields_test.go (hunk 1 above)
Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go]: Thanks. Done :heavy_check_mark:

**Comment by bisakhmondal** [comment]
ci fails to pull docker images
cc @Jaycean, @juzhiyuan, @nic-chen

**Comment by juzhiyuan** [comment]
> ci fails to pull docker images
> cc @Jaycean, @juzhiyuan, @nic-chen

ok, just retrigger

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+
+	ginkgo.It("verify delete route (r1) success", func() {
+		base.RunTestCase(base.HttpTestCase{
+			Object:       base.APISIXExpect(),
+			Method:       http.MethodGet,
+			Path:         "/hello",
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusNotFound,
```
Comment (nic-chen) [inline api/test/e2enew/route/route_with_management_fields_test.go:188]: better to check respond body for each case

Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go:188]: Done :)

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+					"upstream": {
+						"type": "roundrobin",
+						"nodes": {
+							"` + base.UpstreamIp + `:1980": 1
+						}
+					}
+				}`,
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+			Sleep:        time.Duration(2) * time.Second,
```
Comment (nic-chen) [inline api/test/e2enew/route/route_with_management_fields_test.go:142]: need to check body too

Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go:142]: okay. Just for this one or all PUT requests

**Review by Jaycean — APPROVED** [review]
LGTM

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+	)
+
+	ginkgo.It("get the route information", func() {
+		time.Sleep(time.Duration(100) * time.Millisecond)
+		basepath := base.ManagerAPIHost + "/apisix/admin/routes"
+		request, _ := http.NewRequest("GET", basepath+"/r1", nil)
+		request.Header.Add("Authorization", base.GetToken())
+		resp, err := http.DefaultClient.Do(request)
+		if err != nil {
+			return
```
Comment (starsz) [inline api/test/e2enew/route/route_with_management_fields_test.go:103]: Maybe we can use `gomega` to assert the err is nil.

Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go:103]: yeah sure. Good catch.

File: api/test/e2enew/route/route_with_management_fields_test.go (hunk 10 above)
Comment (nic-chen) [reply api/test/e2enew/route/route_with_management_fields_test.go:142]: We can test the fixed part, example:

```go
ExpectBody: []string{`"code":0`, `"name":"new jack"`},
```

File: api/test/e2enew/route/route_with_management_fields_test.go (hunk 11 above)
Comment (starsz) [reply api/test/e2enew/route/route_with_management_fields_test.go]: Well done.Thank you for your contribution.

File: api/test/e2enew/route/route_with_management_fields_test.go [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+			Object:       base.APISIXExpect(),
+			Method:       http.MethodGet,
+			Path:         "/hello",
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+			ExpectBody:   "hello world",
+			Sleep:        base.SleepTime,
+		}),
+		table.Entry("verify the route's detail (r1)", base.HttpTestCase{
+			Object: base.ManagerApiExpect(), ExpectCode: 0,
```
Comment (starsz) [inline api/test/e2enew/route/route_with_management_fields_test.go:233]: The style looks strange.

Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go:233]: Hehe🤡. Mishap from last commit.
Lesson learnt: Never push commits while using public transport 😂😬.
Anyway, fixing asap😅

File: api/test/e2enew/route/route_with_management_fields_test.go (hunk 10 above)
Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go:142]: Cool. Adding changes.

File: api/test/e2enew/route/route_with_management_fields_test.go (hunk 12 above)
Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go]: fixed :)

File: api/test/e2enew/route/route_with_management_fields_test.go (hunk 10 above)
Comment (bisakhmondal) [reply api/test/e2enew/route/route_with_management_fields_test.go]: Now, all tests with PUT request have been updated.

**Review by starsz — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**